### PR TITLE
refactor: add assertion for PrecomputedHasher

### PR DESCRIPTION
**Pull request by fuyufjh**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

`PrecomputedHasher` is somehow hacky and prone to be misused. In a word, `PrecomputedHasher` should ONLY be used along with `HashKey`. If not so, it will silently produce bad hash results.

This PR adds some simple assertions to avoid misusing.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation
None

## Refer to a related PR or issue link (optional)

Come up with it when reviewing https://github.com/risingwavelabs/risingwave/pull/5395#discussion_r980743384

### CODE REVIEW

**Comment by yuhao-su** [comment]
LGTM!

File: src/common/src/hash/key.rs [hunk 1]
```diff
@@ -214,10 +214,13 @@ impl Hash for SerializedKey {
...
-/// A special hasher designed our hashmap, which just stores precomputed hash key.
+/// A special hasher designed for [`HashKey`], which stores a hash key from `HashKey::hash()` and
+/// outputs it on `finish()`.
 ///
 /// We need this because we compute hash keys in vectorized fashion, and we store them in this
 /// hasher.
+///
+/// WARN: This should ONLY be used along with [`HashKey`].
 #[derive(Default)]
 pub struct PrecomputedHasher {
```
Comment (liurenjie1024) [inline src/common/src/hash/key.rs:225]: Is it possible to remove this `pub` modifier and exposes JoinHashMap only?

**Review by liurenjie1024 — APPROVED** [review]

File: src/common/src/hash/key.rs (hunk 1 above)
Comment (fuyufjh) [reply src/common/src/hash/key.rs:225]: Yeah I had thought about it but it's actually used by multiple join executors in both stream/batc

File: src/common/src/hash/key.rs [hunk 2]
```diff
@@ -229,6 +232,8 @@ impl Hasher for PrecomputedHasher {
     }
 
     fn write(&mut self, bytes: &[u8]) {
+        assert_eq!(bytes.len(), 8, "must writes from HashKey with write_u64");
+        assert_eq!(self.hash_code, 0);
         self.hash_code = u64::from_ne_bytes(bytes.try_into().unwrap());
```
[the remark this answers is not in the document]
Comment (fuyufjh) [reply src/common/src/hash/key.rs:237]: Indeed. I'll remove this line 🤔

**Pull request merged**